library/public/types.h

```diff
@@ -82,81 +82,134 @@ class F3D_EXPORT ratio_t
 };
 
 /**
- * Describe a RGB color.
+ * A base template type for an array of double
  */
-class F3D_EXPORT color_t
+template<unsigned int N>
+class double_array_t
 {
 public:
-  color_t() = default;
-  color_t(const std::vector<double>& vec)
+  double_array_t() = default;
+  double_array_t(const std::vector<double>& vec)
   {
-    if (vec.size() != 3)
+    if (vec.size() != N)
     {
       throw f3d::type_construction_exception("Provided vector does not have the right size");
     }
-    std::copy_n(vec.begin(), 3, this->RGB.begin());
-  }
-  color_t(double red, double green, double blue)
-  {
-    this->RGB[0] = red;
-    this->RGB[1] = green;
-    this->RGB[2] = blue;
+    std::copy_n(vec.begin(), N, this->Array.begin());
   }
-  color_t(const std::initializer_list<double>& list)
+  double_array_t(const std::initializer_list<double>& list)
   {
-    if (list.size() != 3)
+    if (list.size() != N)
     {
       throw f3d::type_construction_exception("Provided list does not have the right size");
     }
-    std::copy_n(list.begin(), 3, this->RGB.begin());
+    std::copy_n(list.begin(), N, this->Array.begin());
   }
-  operator std::vector<double>() const
+  [[nodiscard]] operator std::vector<double>() const
   {
-    return std::vector<double>(this->RGB.begin(), this->RGB.end());
+    return std::vector<double>(this->Array.begin(), this->Array.end());
   }
-  bool operator==(const color_t& other) const
+  [[nodiscard]] bool operator==(const double_array_t& other) const
   {
-    return this->RGB == other.RGB;
+    return this->Array == other.Array;
   }
-  bool operator!=(const color_t& other) const
+  [[nodiscard]] bool operator!=(const double_array_t& other) const
   {
-    return this->RGB != other.RGB;
+    return this->Array != other.Array;
   }
-  double operator[](size_t i) const
+  [[nodiscard]] double operator[](size_t i) const
   {
-    if (i >= 3)
+    if (i >= N)
     {
       throw f3d::type_access_exception("Incorrect index");
     }
-    return this->RGB[i];
+    return this->Array[i];
   }
-  double& operator[](size_t i)
+  [[nodiscard]] double& operator[](size_t i)
   {
-    if (i >= 3)
+    if (i >= N)
     {
       throw f3d::type_access_exception("Incorrect index");
     }
-    return this->RGB[i];
+    return this->Array[i];
+  }
+  [[nodiscard]] const double* data() const
+  {
+    return this->Array.data();
+  }
+
+private:
+  std::array<double, N> Array{ 0 };
+};
+
+/**
+ * Describe a RGB color.
+ */
+class color_t : public double_array_t<3>
+{
+public:
+  inline color_t() = default;
+  inline explicit color_t(const std::vector<double>& vec)
+    : double_array_t(vec)
+  {
+  }
+  inline color_t(const std::initializer_list<double>& list)
```

Comment:
@Meakk should that one be explicit too ? Im not sure.

Reply:
I'd say no 🤔